### DISCUSSION TITLE
YAMCS Plugin - Multiple Instance Support

### DISCUSSION THREAD

**Discussion by GUI-Botox** [nasa/openmct#7704] [Community Q&A] [answered]
We're in the process of refactoring the YAMCS plugin to support multiple instances, however, we could use some additional information on the functionality of the user-provider. We've already successfully connected to multiple YAMCS instances to display telemetry and can access the user roles in the first YAMCS instance.

Aside from the exampleUser plugin, we can't find much documentation on the YAMCS integration with poll questions, operator status, and mission status. Although user roles are now selectable once we created them in the first YAMCS instance, mission status isn't populating. Assuming it has something to do with telemetry parameters of types: 'yamcs.operatorStatus', 'yamcs.missionStatus', and 'yamcs.pollQuestion', how should we setup these parameters in YAMCS so that we can complete integration and thoroughly test the providers?

Any thoughts or guidance would be greatly appreciated.

![Multiple YAMCS Instances ](https://github.com/nasa/openmct/assets/129301651/eb0d2d85-fe08-4e14-bd61-958a9694f34a)

**Comment by ozyx** [comment, marked as answer]
Hi @GUI-Botox ,

Apologies for the sparse documentation in this area, and thank you for bringing it to our attention. Here's some [documentation](https://github.com/akhenry/openmct-yamcs/blob/master/src/providers/mission-status/README.md) on how to set up YAMCS MDB and roles to support Mission Status. The process for setting up operator status is very similar to mission status. We use YAMCS aliases to signify which parameters are relevant to which feature.

Below I've provided an example MDB that defines everything needed for the Mission Status and Operator Status features:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<SpaceSystem name="OpenMCTTest" shortDescription="Open MCT Test Parameters"
  xmlns="http://www.omg.org/spec/XTCE/20180204"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xmlns:xi="http://www.w3.org/2001/XInclude">

  <LongDescription>
    Parameters that are used for Open MCT testing only.
  </LongDescription>

  <Header validationStatus="Draft" classification="none" />
  <TelemetryMetaData>
    <ParameterTypeSet>
      <!-- Enumerated parameter type for defining mission status with possible values -->
      <EnumeratedParameterType name="MissionStatus">
        <IntegerDataEncoding sizeInBits="8" encoding="unsigned" />
        <EnumerationList>
          <Enumeration value="0" label="NO GO" shortDescription="No Go" />
          <Enumeration value="1" label="GO" shortDescription="Go" />
        </EnumerationList>
      </EnumeratedParameterType>
      <!-- Enumerated parameter type for defining operator status with possible values -->
      <EnumeratedParameterType name="OperatorStatus">
        <IntegerDataEncoding sizeInBits="8" encoding="unsigned" />
        <EnumerationList>
          <Enumeration value="0" label="NO GO" shortDescription="No Go" />
          <Enumeration value="1" label="GO" shortDescription="Go" />
        </EnumerationList>
      </EnumeratedParameterType>
      <!-- String parameter type for poll questions, allowing for a string of up to 256 bits -->
      <StringParameterType name="PollQuestion">
        <StringDataEncoding sizeInBits="256" />
      </StringParameterType>
    </ParameterTypeSet>
    <!-- Set of parameters referencing the defined parameter types -->
    <ParameterSet>
      <!-- Parameter definition for 'Driving' mission status -->
      <Parameter parameterTypeRef="MissionStatus" name="Driving">
        <AliasSet>
          <!-- Alias defining the parameter type context in OpenMCT -->
          <Alias nameSpace="OpenMCT:type" alias="yamcs.missionStatus" />
          <!-- Alias defining the specific action this parameter represents -->
          <Alias nameSpace="OpenMCT:action" alias="drivingStatus" />
        </AliasSet>
        <ParameterProperties dataSource="local" />
      </Parameter>
      <!-- Parameter definition for 'Imagery' mission status -->
      <Parameter parameterTypeRef="MissionStatus" name="Imagery">
        <AliasSet>
          <Alias nameSpace="OpenMCT:type" alias="yamcs.missionStatus" />
          <Alias nameSpace="OpenMCT:action" alias="imageryStatus" />
        </AliasSet>
        <ParameterProperties dataSource="local" />
      </Parameter>
      <!-- Parameter definition for 'Driver' operator status -->
      <Parameter parameterTypeRef="OperatorStatus" name="Driver">
        <AliasSet>
          <Alias nameSpace="OpenMCT:type" alias="yamcs.operatorStatus" />
          <Alias nameSpace="OpenMCT:role" alias="driverStatus" />
        </AliasSet>
        <ParameterProperties dataSource="local" />
      </Parameter>
      <!-- Parameter definition for 'Flight' operator status -->
      <Parameter parameterTypeRef="OperatorStatus" name="Flight">
        <AliasSet>
          <Alias nameSpace="OpenMCT:type" alias="yamcs.operatorStatus" />
          <Alias nameSpace="OpenMCT:role" alias="flightStatus" />
        </AliasSet>
        <ParameterProperties dataSource="local" />
      </Parameter>
      <!-- Parameter definition for 'Science' operator status -->
      <Parameter parameterTypeRef="OperatorStatus" name="Science">
        <AliasSet>
          <Alias nameSpace="OpenMCT:type" alias="yamcs.operatorStatus" />
          <Alias nameSpace="OpenMCT:role" alias="scienceStatus" />
        </AliasSet>
        <ParameterProperties dataSource="local" />
      </Parameter>
      <!-- Parameter definition for the current poll question -->
      <Parameter parameterTypeRef="PollQuestion" name="CurrentPollQuestion">
        <AliasSet>
          <Alias nameSpace="OpenMCT:type" alias="yamcs.pollQuestion" />
        </AliasSet>
        <ParameterProperties dataSource="local" />
      </Parameter>
    </ParameterSet>
  </TelemetryMetaData>
</SpaceSystem>
```

| Parameter Type  | Encoding Details                     | Aliases and Usage                                                                                   |
|-----------------|--------------------------------------|-----------------------------------------------------------------------------------------------------|
| MissionStatus   | Integer, 8 bits, unsigned            | - Namespace: `OpenMCT:type`, Value: `yamcs.missionStatus`<br>- Example Actions: `Drive`, `Camera` (These are placeholders and can be replaced with actual actions relevant to your mission) |
| OperatorStatus  | Integer, 8 bits, unsigned            | - Namespace: `OpenMCT:type`, Value: `yamcs.operatorStatus`<br>- Example Roles: `Driver`, `Flight`, `Science` (These are placeholders and can be replaced with actual roles relevant to your mission) |
| PollQuestion    | String, 256 bits                     | - Namespace: `OpenMCT:type`, Value: `yamcs.pollQuestion`<br>- Used for current poll questions                           |

### Description
- **MissionStatus**: Enumerated type for mission status values (e.g. "GO", "NO GO").
- **OperatorStatus**: Enumerated type for operator status values (e.g. "NONE", "GO", "NO GO"). 
  - Note: The first status (lowest value enum) will be the default status.
- **PollQuestion**: String type for capturing the current poll question.

### Usage

- **MissionStatus**:
  - **Parameter Type Definition**: Define `MissionStatus` as an enumerated parameter type. This type should enumerate all possible statuses, such as "GO" and "NO GO".
  - **Parameter Instances**: Create individual parameters for each specific mission action. Each parameter should utilize the `MissionStatus` type.
  - **Aliases**:
    - Assign an alias with the namespace `OpenMCT:type` to each parameter instance.
    - The value of the alias should be specific to the mission action it represents (e.g., "Drive" for a driving action).

- **OperatorStatus**:
  - **Parameter Type Definition**: Define `OperatorStatus` as an enumerated parameter type with statuses like "NONE", "GO" and "NO GO".
    - Note: The first (lowest value) enum will be the default status.
  - **Parameter Instances**: Establish parameters for each operator role.
  - **Aliases**:
    - Each parameter should have an alias with the namespace `OpenMCT:type`.
    - The alias value should correspond to the specific operator role (e.g., "Driver" for a driver's status).

### Notes:
- The specific actions and roles mentioned (e.g., "Drive", "Driver") are examples. They can be replaced with anything relevant to your needs.

---

I will update our documentation with the info provided above. Hope this helps!

Cheers,
Jesse

**Reply by GUI-Botox** [reply]
Jesse,

Great info! Thank you for the quick response. We'll reconfigure YAMCS to support Mission Status, Operator Status, and Poll Questions then continue refactoring the Open MCT YAMCS Plugin to support multiple instances.

v/r,
Eric
